README.md
#  Orchard

[![Version](https://img.shields.io/cocoapods/v/Orchard.svg?style=flat)](https://cocoapods.org/pods/Orchard)
[![License](https://img.shields.io/cocoapods/l/Orchard.svg?style=flat)](https://github.com/detroit-labs/Orchard/blob/master/LICENSE)
[![Platform](https://img.shields.io/cocoapods/p/Orchard.svg?style=flat)](https://github.com/detroit-labs/Orchard)

## Device identification for iOS, watchOS, and tvOS.

Orchard is a library for iOS, watchOS, and tvOS to identify the device running
your code. Built in both Swift and Objective-C variants, the code uses the
device model identifier of the system to identify the device.

### Use

_“But I thought we weren’t supposed to do it like this!”_

For most cases, you shouldn’t use the model of the device to control features of
your app. Instead of hardcoding which devices support Touch ID or Face ID, use
`LAContext`. Instead of computing layout based on device model, use Auto Layout
and size classes.

There are certain instances where it _is_ important to know which device your
code is running on. Whether you’re trying to sell accessories for the user’s
current device, [draw onscreen elements in actual size][1], or locate device
hardware relative to the screen, sometimes you need to know the actual device,
and that’s where Orchard comes in.

[1]: https://github.com/detroit-labs/IRLSize

#### Objective-C

To use Orchard in Objective-C, use the `Orchard` category on `UIDevice` (or
`WKInterfaceDevice` on watchOS) to get an enum value representing the
device—either an `OrchardiOSDevice`, `OrchardwatchOSDevice`, or
`OrchardtvOSDevice`, depending on the OS. You can then use this value in a
`switch` statement or pass it to other methods, such as
`OrchardMarketingNameForiOSDevice()`, which returns an `NSString` representing
the marketing name for the device (e.g. “iPhone XS Max”). You can see this in
the example below:

```Objective-C
OrchardiOSDevice device = [UIDevice.currentDevice orchardiOSDevice];

switch (device) {
  case OrchardiOSDeviceiPhoneXSMax:
    NSLog(@"You have an iPhone XS Max!)";
}
```

#### Swift

To use Orchard in Swift, use the `DeviceIdentity` enum. You can obtain the
instance representing the current device using the `deviceIdentity` computed
property of `UIDevice` or `WKInterfaceDevice`. Instances of the `DeviceIdentity`
enum have a `marketingName` property that returns the marketing name for the
device (e.g. “iPad Pro (12.9-inch) (3rd Generation)”). You can see this in the
example below:

```Swift
let deviceIdentity = UIDevice.current.deviceIdentity

print("I’m an \(deviceIdentity.marketingName ?? "unknown device").")

switch deviceIdentity {
case .iPhone(.iPhone11ProMax), .iPhone(.iPhoneXSMax), .iPhone(.iPhone8Plus),
     .iPhone(.iPhone7Plus), .iPhone(.iPhone6sPlus), .iPhone(.iPhone6Plus),
     .iPad(.iPadPro12_9Inch), .iPad(.iPadPro12_9Inch2),
     .iPad(.iPadPro12_9Inch3), .iPad(.iPadPro12_9Inch4):
    print("You got the big one!")
default: break
}
```

#### Mixed-Source Projects

Orchard’s Objective-C code is completely usable from Swift; device names are
annotated with [`NS_SWIFT_NAME`][2] and therefore can be used in the above
example just as their Swift-native counterparts.

[2]: https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/renaming_objective-c_apis_for_swift

### Installation

#### CocoaPods

Orchard can be installed using CocoaPods. There are two subspecs, `Swift` and
`ObjC`, so use either of these lines in your `Podfile`:

`pod "Orchard/ObjC"`

`pod "Orchard/Swift"`

#### Swift Package Manager

The Swift portion of Orchard can be installed using the
[Swift package manager][3].

[3]: https://swift.org/package-manager/

### Development

Orchard uses [Swift gyb] to generate both Objective-C and Swift source files.
The actual device data is in the `Data` folder, in TSV format. To add new
devices, simply update the TSV file(s) with new device info and rebuild both
targets; the source code will be regenerated. Do not modify any file for which
there is a corresponding `.gyb` file, as the builds will overwrite your changes.

To update the generated source files, run the following command:

```Shell
make generate
```


CHANGELOG.md
# Orchard Change Log

## [1.15.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.15.0)
**Released:** 2023-08-30

### Updated
- Set minimum iOS deployment target to 12.0
- Set minimum tvOS deployment target to 12.0
- Set minimum watchOS deployment target to 4.0

## [1.14.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.14.0)
**Released:** 2023-08-30

### Added
- Support for Apple TV 4K (3rd Generation)

### Updated
- Support for Apple TV 4K (2nd Generation)

## [1.13.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.13.0)
**Released:** 2022-10-05

### Added
- Support for iPhone 14, iPhone Plus, iPhone 14 Pro, iPhone 14 Pro Max
- Support for Apple Watch SE (2nd Generation)
- Support for Apple Watch Series 8
- Support for Apple Watch Ultra

## [1.12.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.12.0)
**Released:** 2022-05-23

### Added
- Support for iPhone SE (3rd Generation)
- Support for iPad Air (5th Generation)
- Support for Apple TV 4K (2nd Generation)

## [1.11.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.11.0)
**Released:** 2021-10-18

### Added
- Support for iPhone 13, iPhone 13 mini, iPhone 13 Pro, iPhone 13 Pro Max.
- Support for Apple Watch Series 7
- Support for iPad (9th Generation)
- Support for iPad mini (6th Generation)
- Support for iPad Pro (12.9-inch), iPad Pro (11-inch)

## [1.10.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.10.0)
**Released:** 2021-01-12

### Added
- Support for Apple Silicon

## [1.9.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.9.0)
**Released:** 2021-01-08

### Updated
- Moved to the [Detroit Labs](https://github.com/Detroit-Labs) organization.

## [1.8.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.8.0)
**Released:** 2020-10-15

### Added
- Support for iPhone 12, iPhone 12 mini, iPhone 12 Pro, and iPhone 12 Pro Max.

### Updated
- iOS deployment target set to 9.0 to match Xcode 12’s support.
- Reordered iPad models to group by device type.

## [1.7.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.7.1)
**Released:** 2020-10-09

### Updated
- Additional identifier for iPad Air (4th Generation).

## [1.7.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.7.0)
**Released:** 2020-09-17

### Added
- Support for Apple Watch SE and Apple Watch Series 6.
- Support for iPad (8th Generation) and iPad Air (4th Generation).

## [1.6.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.6.1)
**Released:** 2020-08-05

### Added
- Deprecation versions for devices deprecated by watchOS 7.

## [1.6.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.6.0)
**Released:** 2020-04-15

### Added
- Support for iPhone SE (2nd Generation).

## [1.5.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.5.1)
**Released:** 2020-03-39

### Added
- Support for homebrew-bundle.

### Updated
- README instructions now include homebrew-bundle.

### Fixed
- README instructions now work on a fresh macOS install.
- All Markdown files wrap at 80 characters for more pleasant viewing in the
  default macOS Terminal.

## [1.5.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.5.0)
**Released:** 2020-03-24

### Added
- Support for iPad Pro (11-inch) (2nd Generation) and iPad Pro (12.9-inch)
(4th Generation).

### Updated
- iPad Pro marketing names updated to more closely match Apple terminology.

## [1.4.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.4.1)
**Released:** 2020-03-10

### Added
- PR build GitHub Action.

### Modified
- Extract simulator identifier into category.

## [1.4.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.4.0)
**Released:** 2020-02-25

### Added
- Support for Swift Package Manager.

## [1.3.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.3.0)
**Released:** 2019-09-24

### Added
- Support for iPad (7th Generation).

## [1.2.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.2.1)
**Released:** 2019-09-22

### Added
- Support for Swift 5.1 in the Podspec for CocoaPods.

## [1.2.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.2.0)
**Released:** 2019-09-16

### Added
- Support for iPhone 11, iPhone 11 Pro, iPhone 11 Pro Max, and Apple Watch
Series 5.

## [1.1.3](https://github.com/Detroit-Labs/Orchard/releases/tag/1.1.3)
**Released:** 2019-09-04

### Fixed
- Fixed typo in Objective-C enum name.

## [1.1.2](https://github.com/Detroit-Labs/Orchard/releases/tag/1.1.2)
**Released:** 2019-06-16

### Fixed
- Fixed typo in iPad 4 identifier.

## [1.1.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.1.1)
**Released:** 2019-06-03
 
### Updated
 - Added deprecation information for iOS and iPadOS 13.

## [1.1.0](https://github.com/Detroit-Labs/Orchard/releases/tag/1.1.0)
**Released:** 2019-05-29

### Added
 - Added iPod touch (7th Generation).
  
## [1.0.1](https://github.com/Detroit-Labs/Orchard/releases/tag/1.0.1)
**Released:** 2019-05-22

### Updated
  - Podspec syntax updated to CocoaPods 1.7 syntax
  - Supported Swift versions set to 4.0 - 5.0.
  - Added `CHANGELOG.md`.


Common/GYBUtilities.py
import csv
from enum import Enum
import os

class OSFamily(Enum):
	IOS = "iOS"
	WATCHOS = "watchOS"
	TVOS = "tvOS"

class iOSDeviceFamily(Enum):
	IPHONE = "iPhone"
	IPAD = "iPad"
	IPOD = "iPod"

class watchOSDeviceFamily(Enum):
	WATCH = "Watch"

class tvOSDeviceFamily(Enum):
	TV = "AppleTV"

def DataPath(deviceFamily = Enum):
	if deviceFamily is iOSDeviceFamily.IPHONE:
		return '../../Data/iPhones.tsv'
	elif deviceFamily is iOSDeviceFamily.IPAD:
		return '../../Data/iPads.tsv'
	elif deviceFamily is iOSDeviceFamily.IPOD:
		return '../../Data/iPods.tsv'
	elif deviceFamily is watchOSDeviceFamily.WATCH:
		return '../../Data/Watches.tsv'
	elif deviceFamily is tvOSDeviceFamily.TV:
		return '../../Data/TVs.tsv'

def longestEnumNameLength(deviceFamily = iOSDeviceFamily.IPHONE):
	longest = 0
	with open(DataPath(deviceFamily)) as file:
		for row in csv.DictReader(file, delimiter='\t'):
			length = len(enumName(row["Name"]))
			if length > longest:
				longest = length
	return longest

iOSDeploymentTarget=os.getenv('IPHONEOS_DEPLOYMENT_TARGET', '12.0')
watchOSDeploymentTarget=os.getenv('WATCHOS_DEPLOYMENT_TARGET', '4.0')
tvOSDeploymentTarget=os.getenv('TVOS_DEPLOYMENT_TARGET', '12.0')

def versiontuple(v):
	return tuple(map(int, v.split('.')))

def versiontupleAdd(u, v):
	return tuple(x + y for x, y in zip(versiontuple(u), versiontuple(v)))

def versiontupleToString(v):
	return ".".join(map(str, v))

def canSupport(version, os = OSFamily.IOS):
	minimum = versiontuple(iOSDeploymentTarget)

	if os == OSFamily.WATCHOS:
		minimum = versiontuple(watchOSDeploymentTarget)
	elif os == OSFamily.TVOS:
		minimum = versiontuple(tvOSDeploymentTarget)

	return versiontuple(version) > minimum

def needsAvailabilityCheck(version, os = OSFamily.IOS):
	minimum = versiontuple(iOSDeploymentTarget)
	
	if os == OSFamily.WATCHOS:
		minimum = versiontuple(watchOSDeploymentTarget)
	elif os == OSFamily.TVOS:
		minimum = versiontuple(tvOSDeploymentTarget)
	
	return versiontuple(version) > minimum

def enumName(x):
	return (x
		.replace(" ", "")
		.replace("(", "").replace(")","")
		.replace(".", "_").replace("-", "").replace(",", "")
		.replace("mini", "Mini")
		.replace("Generation", "")
		.replace("inch", "Inch")
		.replace("1st", "").replace("2nd", "2").replace("3rd", "3")
		.replace("4th", "4").replace("5th", "5").replace("6th", "6")
		.replace("7th", "7").replace("8th", "8").replace("9th", "9")
		.replace("touch", "Touch")
	)
